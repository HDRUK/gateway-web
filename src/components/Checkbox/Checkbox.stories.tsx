import React from "react";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { Stack } from "@mui/material";
import type { Meta, StoryObj } from "@storybook/react";
import * as yup from "yup";
import Button from "@/components/Button";
import CheckboxComponent from "@/components/Checkbox";
import Form from "@/components/Form";

/** Mui documentation: https://mui.com/material-ui/react-checkbox/ */
/** React Hook Form documentation: https://react-hook-form.com/ */

const meta: Meta<typeof Form> = {
    component: Form,
    title: "Forms/Checkbox",
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Form>;

export type FormData = {
    first: string;
    second: boolean;
    third: string;
    fourth: string;
};

const validationSchema = yup
    .object({
        second: yup.boolean().required().oneOf([true]),
    })
    .required();

const WrapperComponent = () => {
    const { handleSubmit, control } = useForm<FormData>({
        defaultValues: {
            first: "",
            second: false,
            third: "",
            fourth: "",
        },
        resolver: yupResolver(validationSchema),
    });

    const onSubmit = (data: unknown) => console.log(data);

    return (
        <Form onSubmit={handleSubmit(onSubmit)}>
            <Stack spacing={2} sx={{ marginBottom: 4, maxWidth: 240 }}>
                <CheckboxComponent
                    label="Simple Checkbox"
                    control={control}
                    name="first"
                />
                <CheckboxComponent
                    label="Required Checkbox"
                    required
                    control={control}
                    name="second"
                />
                <CheckboxComponent
                    label="Disabled Checkbox"
                    disabled
                    control={control}
                    name="third"
                />
                <CheckboxComponent
                    label="indeterminate Checkbox"
                    indeterminate
                    control={control}
                    name="fourth"
                />
                <Button type="submit">Submit</Button>
            </Stack>
        </Form>
    );
};

const WrapperComponentWithCount = () => {
    const { handleSubmit, control } = useForm<FormData>({
        defaultValues: {
            first: "",
            second: false,
            third: "",
            fourth: "",
        },
        resolver: yupResolver(validationSchema),
    });

    const onSubmit = (data: unknown) => console.log(data);

    return (
        <Form onSubmit={handleSubmit(onSubmit)}>
            <Stack spacing={2} sx={{ marginBottom: 4, maxWidth: 240 }}>
                <CheckboxComponent
                    label="Simple Checkbox"
                    control={control}
                    count={34}
                    name="first"
                />
                <CheckboxComponent
                    label="Required Checkbox"
                    required
                    control={control}
                    count={23}
                    name="second"
                />
                <CheckboxComponent
                    label="Disabled Checkbox"
                    disabled
                    control={control}
                    count={2}
                    name="third"
                />
                <CheckboxComponent
                    label="indeterminate Checkbox"
                    indeterminate
                    control={control}
                    count={12}
                    name="fourth"
                />
                <Button type="submit">Submit</Button>
            </Stack>
        </Form>
    );
};

export const Checkbox: Story = {
    render: () => <WrapperComponent />,
};

export const WithCount: Story = {
    render: () => <WrapperComponentWithCount />,
};
