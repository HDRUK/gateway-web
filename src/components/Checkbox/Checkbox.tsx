/** @jsxImportSource @emotion/react */

import MuiCheckbox, {
    CheckboxProps as MuiCheckboxProps,
} from "@mui/material/Checkbox";
import { FormControl, FormControlLabel, SxProps } from "@mui/material";
import { Control, useController } from "react-hook-form";
import {
    CheckBoxOutlineBlankSharpIcon,
    CheckBoxSharpIcon,
    IndeterminateCheckBoxSharpIcon,
} from "@/consts/icons";
import FormError from "@/components/FormError";

export interface CheckboxProps extends MuiCheckboxProps {
    label: string;
    name: string;
    fullWidth: boolean;
    control: Control;
    checkboxSx?: SxProps;
    formControlSx?: SxProps;
}

const Checkbox = (props: CheckboxProps) => {
    const {
        fullWidth = true,
        label,
        control,
        name,
        checkboxSx,
        formControlSx,
        ...rest
    } = props;

    const {
        field: { ref, ...fieldProps },
        fieldState: { error },
    } = useController({
        name,
        control,
    });

    return (
        <FormControl
            fullWidth={fullWidth}
            sx={{ m: 0, mb: 2, ...formControlSx }}
            error={!!error}>
            <FormControlLabel
                control={
                    <MuiCheckbox
                        color={error !== undefined ? "error" : "secondary"}
                        disableRipple
                        checked={fieldProps.value}
                        size="medium"
                        icon={<CheckBoxOutlineBlankSharpIcon />}
                        checkedIcon={<CheckBoxSharpIcon />}
                        indeterminateIcon={<IndeterminateCheckBoxSharpIcon />}
                        inputRef={ref}
                        sx={{ ...checkboxSx }}
                        {...rest}
                        {...fieldProps}
                    />
                }
                label={label}
            />
            {error && <FormError error={error} />}
        </FormControl>
    );
};

Checkbox.defaultProps = {
    checkboxSx: {},
    formControlSx: {},
};

export default Checkbox;
