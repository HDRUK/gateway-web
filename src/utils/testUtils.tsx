import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
