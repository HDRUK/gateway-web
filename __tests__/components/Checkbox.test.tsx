import React from "react";
import Checkbox from "@/components/Checkbox";
import { useForm } from "react-hook-form";
import { render, screen } from "../testUtils";

describe("Checkbox", () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Component = (props: any) => {
        const { control } = useForm();
        return (
            <Checkbox
                label="checkbox label"
                name="fieldName"
                control={control}
                {...props}
            />
        );
    };
    it("should render component", async () => {
        render(<Component />);

        expect(screen.getByText("checkbox label")).toBeInTheDocument();
    });

    it("checkbox snapshot", async () => {
        const wrapper = render(<Component />);

        expect(wrapper.container).toMatchSnapshot();
    });
});
