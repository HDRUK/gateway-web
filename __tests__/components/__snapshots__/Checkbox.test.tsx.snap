// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox checkbox snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth mui-style-vd0s50-MuiFormControl-root"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd mui-style-1pwbvyj-MuiFormControlLabel-root"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-root MuiCheckbox-colorSecondary mui-style-dzoou-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          class="PrivateSwitchBase-input mui-style-1m9pwf3"
          data-indeterminate="false"
          name="fieldName"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-style-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label mui-style-ahj2mt-MuiTypography-root"
      >
        checkbox label
      </span>
    </label>
  </div>
</div>
`;
